numpy/random/tests/test_random.py

```diff
@@ -209,6 +243,38 @@ def test_repeatability(self):
         res = hashlib.md5(val).hexdigest()
         assert_(tgt[np.dtype(np.bool).name] == res)
 
+    def test_int64_uint64_corner_case(self):
+        # When stored in Numpy arrays, `lbnd` is casted
+        # as np.int64, and `ubnd` is casted as np.uint64.
+        # Checking whether `lbnd` >= `ubnd` used to be
+        # done solely via direct comparison, which is incorrect
+        # because when Numpy tries to compare both numbers,
+        # it casts both to np.float64 because there is
+        # no integer superset of np.int64 and np.uint64. However,
+        # `ubnd` is too large to be represented in np.float64,
+        # causing it be round down to np.iinfo(np.int64).max,
+        # leading to a ValueError because `lbnd` now equals
+        # the new `ubnd`.
+
+        dt = np.int64
+        tgt = np.iinfo(np.int64).max
+        lbnd = np.int64(np.iinfo(np.int64).max)
+        ubnd = np.uint64(np.iinfo(np.int64).max + 1)
+
+        # None of these function calls should
+        # generate a ValueError now
+        actual = np.random.randint(lbnd, ubnd, dtype=dt)
+        assert_equal(actual, tgt)
+
+        actual = np.random.randint([lbnd] * 3, ubnd, dtype=dt)
+        assert_equal(actual, tgt)
+
+        actual = np.random.randint(lbnd, [ubnd] * 3, dtype=dt)
+        assert_equal(actual, tgt)
+
+        actual = np.random.randint([lbnd] * 3, [ubnd] * 3, dtype=dt)
```

Comment:
> but I think disagreeing about how to handle large ints is not something that should be hidden away in the diff comments for a PR titled "broadcasting" ;). 

Ah, but handling large `int` was a big issue in the implementation of broadcasting.  That's why there are entire tests dedicated to it.  But if you do say that the scalar version of the tests fail on `master`, then sure I can break it off this PR to simplify this one.

> Because it comes for free as a result of unsigned integer wraparound

Just because it comes for free, doesn't mean we should expose it in the API.  It makes the purpose of this function less intuitive in my opinion because then the "bounds" on each `dtype` cannot be enforced as strictly as I would want it to be.

> I'm not the first to propose this

True, but we ultimately merged this PR (#6931) in shortly after that comment was made.